### PR TITLE
WIP: [JENKINS-48822] Better browser URL guessing

### PR DESCRIPTION
**Pull request by Gyanesha**
# [JENKINS-48822](https://issues.jenkins-ci.org/browse/JENKINS-48822) - Improve browser URL guessing

I have made the abstract BrowserInference Class which is the extension point and can be used by other GitRepository Browsers.

## This is a work in progress. Do not merge

### CODE REVIEW

**Comment by MarkEWaite** [comment]
I had suggested that the steps would be:

* Add an extension (done in this pull request, not sure on the naming choice)
* Write tests for that extension to confirm it behaves as initially implemented
* Implement a method in the extension which takes the repository URL as an argument and returns either a GitRepositoryBrowser on a match or null
* Write tests of the new method to confirm matching and non-matching cases
* Extend the GitSCM constructor to iterate over all implementations of that extension, calling each extension with the repository URL, accepting the first GitRepositoryBrowser returned as the default
* Implement the extension in one of the well-known existing repository browsers (GitHub, Bitbucket, GitLab, or Gitea)
* Write tests of the extension implementation in that repository browser
* Confirm the extension implementation in that repository browser works

File: src/main/java/hudson/plugins/git/extensions/BrowserInference.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
+package hudson.plugins.git.extensions;
+
+import hudson.plugins.git.browser.GitRepositoryBrowser;
+import hudson.scm.RepositoryBrowsers;
+import org.apache.tools.ant.ExtensionPoint;
```
Comment (fcojfernandez) [inline src/main/java/hudson/plugins/git/extensions/BrowserInference.java]: As https://wiki.jenkins.io/display/JENKINS/Defining+a+new+extension+point mention, you're using the wrong extension point class
```suggestion
import hudson.ExtensionPoint;
```

File: src/main/java/hudson/plugins/git/extensions/BrowserInference.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+public abstract class BrowserInference extends ExtensionPoint {
```
Comment (fcojfernandez) [inline src/main/java/hudson/plugins/git/extensions/BrowserInference.java]: Here you should wonder:
1. Why abstract?
1. Will you allow the rest of plugins to consume this extension? Or is it something that will be part only of the git plugin?
1. Is some sort of UI needed? Will you offer a new UI?

If this extension is something that will never extended by other plugins, then:
1. Annotate it as `@Restricted(NoExternalUse.class)` as in https://github.com/jenkinsci/git-plugin/blob/c0e8144da63e3c8ccea7f00f7438b6e70aac9b00/src/main/java/jenkins/plugins/git/traits/GitBrowserSCMSourceTrait.java#L110
1. There is no need to make it abstract. Use abstract classes when you want them to be extended.

If there is some expected offered UI, you should implement the Descriptor. See the `Describable/Descriptor pattern` in  https://wiki.jenkins.io/display/JENKINS/Defining+a+new+extension+point

File: src/main/java/hudson/plugins/git/extensions/BrowserInference.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+        browser = null;
+    }
+
+    public BrowserInference(GitRepositoryBrowser browser){
+        this.browser = browser;
+    }
+
+    String inferRepositoryBrowser(){
+            return "no-browser";
+    }
```
Comment (fcojfernandez) [inline src/main/java/hudson/plugins/git/extensions/BrowserInference.java]: I might be misunderstanding Mark's suggestion, but from his words my understanding is exactly the opossite:
>  I assume that the extension point would need a method that would be passed the repository URL and would either return an instance of GitRepositoryBrowser or null. The null return would mean that the extension rejected the request to provide a browser, while if the GitRepositoryBrowser was returned, then that would be the inferred browser.

You have to receive a String/URL object and return the proper GitRepositoryBrowser. The place for the logic to calculate/detect the URL is there. Returning the GitRepositoryBrowser you will have the same effect as when you select a browser. At this point, have you run a Jenkins with the plugin installed and have you seen the current behaviour? So I can sure we are on the same page and we're understanding each other.

**Comment by MarkEWaite** [comment]
Closing this until someone decides to make further progress on it.

**Comment by MarkEWaite** [comment]
> Hey! I was looking into this issue.
> 
> > 1. Add an extension (done in this pull request, not sure on the naming choice)
> > 2. Write tests for that extension to confirm it behaves as initially implemented
> > 3. Implement a method in the extension which takes the repository URL as an argument and returns either a GitRepositoryBrowser on a match or null
> > 4. Write tests of the new method to confirm matching and non-matching cases
> > 5. Extend the GitSCM constructor to iterate over all implementations of that extension, calling each extension with the repository URL, accepting the first GitRepositoryBrowser returned as the default
> > 6. Implement the extension in one of the well-known existing repository browsers (GitHub, Bitbucket, GitLab, or Gitea)
> > 7. Write tests of the extension implementation in that repository browser
> > 8. Confirm the extension implementation in that repository browser works
> 
> In the above points, I didn't understand the 5th point. Can someone explain me the 5th point.

A new extension point will allow other plugins (like the Gitea plugin or the Tuleap branch source plugin or the Gerrit plugin) to provide their own implementation without modifying the git plugin.  There are a few examples in the GitSCM class where the code iterates over all available extensions (`for (GitSCMExtension ext : getExtensions())`) and uses a result from that iteration.

I think that was the intent of that description

**Pull request closed without merging**